### PR TITLE
System tray

### PR DESCRIPTION
**Pull request by neuroblack**
Adds start and exit with system tray behaviours.
Both can be toggled on demand in general settings. There is some new entries in the i18n dictionary.

FYI : For now the splash screen is not hidden when start with system tray is enabled, only the main window is.

### CODE REVIEW

File: src/app/models/settings.ts [hunk 1]
```diff
@@ -9,6 +9,8 @@ export interface AppSettings {
   oscReceivingPort: number;
   enableDesktopNotifications: boolean;
   enableXSOverlayNotifications: boolean;
+  exitInSystemTray: boolean;
+  startInSystemTray: boolean;
 }
```
Comment (Raphiiko) [inline src/app/models/settings.ts:14]: So these interfaces have a `version` property that gets bumped whenever they are modified. This is so that people's existing configurations can be migrated once they update. 

In this case, two new fields got added, which means the `version` property should be bumped from `4` to `5`, and a migration should be written for it in `src/app/migrations/app-settings.migrations.ts`. If you take a look at the previous migrations in there, you'll see it's quite simple!

File: src/app/services/system-tray.service.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import { Injectable } from '@angular/core';
+import { AppSettingsService } from './app-settings.service';
+import { debounceTime, filter, first, last, map, pairwise, take } from 'rxjs';
```
Comment (Raphiiko) [inline src/app/services/system-tray.service.ts:3]: Nit: there's still a few unused imports here :)

File: src/app/services/system-tray.service.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  }
+  
+  async init() {
+    // Update exit in system tray behaviour following the setting.
+    this._appSettingsService.settings
+      .pipe(
+        map(settings => settings.exitInSystemTray),
+        pairwise(),
+        filter(([oldVal, newVal]) => oldVal !== newVal),
+        map(([_, newVal]) => newVal),
```
Comment (Raphiiko) [inline src/app/services/system-tray.service.ts:23]:
```suggestion
        map(([, newVal]) => newVal),
```
Nit: in the case of unused array parameters, they can be left empty!

File: src/app/services/system-tray.service.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      .subscribe(exitInSystemTray => this.updateSystemTrayExit(exitInSystemTray));
+
+    // Update start in system tray behaviour following the settings.
+    // Send command only upon loading setings, in order to hide or show the window upon startup.
+    this._appSettingsService.settings
+      .pipe(
+        map(settings => settings.startInSystemTray),
+        pairwise(),
+        filter(([oldVal, newVal]) => oldVal !== newVal),
+        map(([_, newVal]) => newVal),
```
Comment (Raphiiko) [inline src/app/services/system-tray.service.ts:35]:
```suggestion
        map(([, newVal]) => newVal),
```

File: src/app/views/dashboard-view/views/settings-view/settings-general-tab/settings-general-tab.component.html [hunk 5]
```diff
@@ -74,6 +74,46 @@ <h2 translate>settings.general.adminPrivileges.title</h2>
     </div>
   </div>
 </div>
+
+<div class="setting-category">
+  <h2 translate>settings.general.systemTray.title</h2>
+  <div class="settings">
+    <div class="setting-row">
+      <div class="setting-row-label" translate>
```
Comment (Raphiiko) [inline src/app/views/dashboard-view/views/settings-view/settings-general-tab/settings-general-tab.component.html:82]:
```suggestion
      <div class="setting-row-label">
```
Nit: unneeded translate directive

File: src/app/views/dashboard-view/views/settings-view/settings-general-tab/settings-general-tab.component.html [hunk 6]
```diff
@@ -74,6 +74,46 @@ <h2 translate>settings.general.adminPrivileges.title</h2>
...
+            type="checkbox"
+            [checked]="appSettings.exitInSystemTray"
+            (change)="setExitInSystemTray(!appSettings.exitInSystemTray)"
+          />
+          <span class="switch-toggle-slider"></span>
+        </label>
+      </div>
+    </div>
+    <div class="setting-row">
+      <div class="setting-row-label" translate>
```
Comment (Raphiiko) [inline src/app/views/dashboard-view/views/settings-view/settings-general-tab/settings-general-tab.component.html:98]:
```suggestion
      <div class="setting-row-label">
```

File: src/assets/i18n/en.json [hunk 7]
```diff
@@ -424,6 +424,17 @@
         "title": "Telemetry",
         "label": "Send anonymous usage data",
         "description": "The data that will be sent is limited to the version of Oyasumi that you are using, and the language you have set. No other data is sent and it is completely anonymous. This data is only used for providing some general usage statistics."
+      },
+      "systemTray": {
+        "title": "System tray",
+        "close": {
+          "label": "Exit in system tray",
```
Comment (Raphiiko) [inline src/assets/i18n/en.json:431]: Imo a bit more clear, as it doesn't actually exit
```suggestion
          "label": "Close to system tray",
```

File: src/assets/i18n/en.json [hunk 8]
```diff
@@ -424,6 +424,17 @@
...
+          "description": "Collapses the program into the system tray upon clicking on exit button"
```
Comment (Raphiiko) [inline src/assets/i18n/en.json:432]:
```suggestion
          "description": "When closing Oyasumi, it will continue to run in the system tray"
```

File: src/assets/i18n/en.json [hunk 9]
```diff
@@ -424,6 +424,17 @@
...
+        },
+        "start": {
+          "label": "Start in system tray",
+          "description": "Starts the program reduced in the system tray."
```
Comment (Raphiiko) [inline src/assets/i18n/en.json:436]:
```suggestion
          "description": "The main window will remain hidden when Oyasumi is started"
```

File: src/assets/i18n/fr.json [hunk 10]
```diff
@@ -417,6 +417,17 @@
...
+      "systemTray": {
+        "title": "Barre d'état",
+        "close": {
+          "label": "Fermer l'application dans la barre d'état",
+          "description": "Réduit l'application dans la barre d'état système lorsque vous cliquez sur le bouton 'Fermer'"
+        },
+        "start": {
+          "label": "Démarrer dans la barre d'état",
+          "description": "Démarrer l'application cachée dans la barre d'état. Vous pouvez la restaurer en cliquant sur l'icône dans la barre d'état."
+        }
```
Comment (Raphiiko) [inline src/assets/i18n/fr.json:430]: Maybe these need to be updated based on the previous changes to the English translations?

File: src-tauri/src/commands/system_tray.rs [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    //         }
+    //         _ => {}
+    //     }
+    // }
+}
+
+// Initializes the system tray with menus.
+pub fn init_system_tray() -> SystemTray {
+    // Menus
+    let menu_quit = CustomMenuItem::new(QUIT, "Quit");
```
Comment (Raphiiko) [inline src-tauri/src/commands/system_tray.rs:35]: We should probably find a way of localizing this as well, but I understand there's currently not yet a nice way of getting the translations to the rust side. I think doing that might be a bit outside of this PR's scope, especially when you consider the translations should change when the user switches language on the front, so that's probably something to look into after getting this merged.

tl;dr: this is fine as is, feel free to ignore. I've opened a separate issue for this in #23.

File: src-tauri/src/commands/system_tray.rs [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    
+    // pub fn handle_window_exit<R: Runtime>(&self) -> impl Fn(GlobalWindowEvent<R>) + Send + Sync + 'static {
+    //     return |event| match event.event() {
+    //         tauri::WindowEvent::CloseRequested { api, .. } => {
+    //             event.window().hide().unwrap();
+    //             api.prevent_close();
+    //         }
+    //         _ => {}
+    //     }
+    // }
```
Comment (Raphiiko) [inline src-tauri/src/commands/system_tray.rs:29]: Since it's commented out, I assume this can be removed?

**Review by Raphiiko — CHANGES_REQUESTED** [review]
Thanks a bunch for the contribution! It's looking great, the code you wrote is really clean ^^ 

I have some comments on it below, they're mostly just nits and one missing part for the app config migration. Would you be able to update those last few things? Once those changes are in I'm pretty sure we can get this merged.

File: src/app/models/settings.ts [hunk 13]
```diff
@@ -9,6 +9,8 @@ export interface AppSettings {
   oscReceivingPort: number;
   enableDesktopNotifications: boolean;
   enableXSOverlayNotifications: boolean;
+  exitInSystemTray: boolean;
```
Comment (Raphiiko) [inline src/app/models/settings.ts:12]: Can we refactor this to `closeToSystemTray`?

File: src-tauri/src/commands/system_tray.rs [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    AppHandle, CustomMenuItem, Runtime, SystemTray, SystemTrayEvent, SystemTrayMenu,
+};
+
+use crate::SYSTEMTRAY_MANAGER;
+
+const QUIT: &'static str = "quit";
+
+#[derive(Debug, Clone)]
+pub struct SystemTrayManager {
+    pub exit_in_tray: bool,
```
Comment (Raphiiko) [inline src-tauri/src/commands/system_tray.rs:12]: Same here I guess
```suggestion
    pub close_to_tray: bool,
```

**Comment by Raphiiko** [comment]
As I've refactored a majority of Oyasumi's Rust core, this PR ended up requiring quite a bit more changes to fit into the new structure. I didn't really want to ask to update your work to get it working again, so I've made a few changes for it myself so that it could be merged. You'll find the changes in 7776f991008bdc38900ef9bcfc0f988353441a53. Thanks a bunch for your contribution!
